### ISSUE TITLE
DataViews component

### ISSUE THREAD

**Issue by youknowriad** [WordPress/gutenberg#55083]
This document keeps track of all the work and tasks to build a powerful DataViews component capable of rendering all kind of lists of data in WordPress Admin and beyond: from Post Lists, taxonomies, patterns, templates...

DataViews is based on a Fields abstraction and support searching, filtering, reordering fields, multiple layouts, pagination and more.

<img width="1132" height="804" alt="Image" src="https://github.com/user-attachments/assets/c74bf6b6-17c0-4cae-8400-d48e54d55022" />

> See also issues and PRs tagged with the [`[Feature] Data Views`](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Data%20Views) label.

## DataViews component

 - [x] Support defining actions per row. [#55096](https://github.com/WordPress/gutenberg/issues/55096)
 - [x] Add support for reordering fields. [#63414](https://github.com/WordPress/gutenberg/issues/63414)
 - [x] Implement a new layout (or view type): Grid/Gallery.
 - [x] Iterations on the grid layout: [#56643](https://github.com/WordPress/gutenberg/issues/56643)
 - [x] Implement the list layout. [#56746](https://github.com/WordPress/gutenberg/issues/56746)
 - [x] List View Iterations [#56983](https://github.com/WordPress/gutenberg/issues/56983) [#57116](https://github.com/WordPress/gutenberg/issues/57116) [#57578](https://github.com/WordPress/gutenberg/issues/57578) [#57523](#57523) [#56602](https://github.com/WordPress/gutenberg/issues/56602) [#57649](https://github.com/WordPress/gutenberg/issues/57649) [#56746](https://github.com/WordPress/gutenberg/issues/56746)
 - [x] Keyboard navigation: potentially use arrow up/down to navigate through the list.  [#57895](https://github.com/WordPress/gutenberg/issues/57895) [#59637](https://github.com/WordPress/gutenberg/issues/59637)
 - [x] Iterations on search and filtering [#55905](https://github.com/WordPress/gutenberg/issues/55905) [#60696](https://github.com/WordPress/gutenberg/issues/60696) [#63152](https://github.com/WordPress/gutenberg/issues/63152)
 - [x] Bulk actions [#55098](https://github.com/WordPress/gutenberg/issues/55098) [#63193](https://github.com/WordPress/gutenberg/issues/63193) [#63150](https://github.com/WordPress/gutenberg/issues/63150)
 - [x] Field registration API
 - [x] Pagination [#55095](https://github.com/WordPress/gutenberg/issues/55095) [#55937](https://github.com/WordPress/gutenberg/issues/55937) [#55164](https://github.com/WordPress/gutenberg/issues/55164) [#63675](https://github.com/WordPress/gutenberg/issues/63675)
 - [x] View config iterations [#63195](https://github.com/WordPress/gutenberg/issues/63195) [#63872](https://github.com/WordPress/gutenberg/issues/63872)
 - [x] Implement custom views.
 - [x] Support density [#60630](https://github.com/WordPress/gutenberg/issues/60630) [#66312](https://github.com/WordPress/gutenberg/issues/66312)
 - [x] Support different fields formats [#60284](https://github.com/WordPress/gutenberg/issues/60284) [#60083 ](https://github.com/WordPress/gutenberg/issues/60083 ) 
 - [x] Accessibility iterations [#56396](https://github.com/WordPress/gutenberg/issues/56396) [#56853](https://github.com/WordPress/gutenberg/issues/56853) [#56441](https://github.com/WordPress/gutenberg/issues/56441) [#57341](https://github.com/WordPress/gutenberg/issues/57341) [#56001](https://github.com/WordPress/gutenberg/issues/56001) [#57978](https://github.com/WordPress/gutenberg/issues/57978) [#58977](https://github.com/WordPress/gutenberg/issues/58977)
 - [x] Some Bug fixes [#56154](https://github.com/WordPress/gutenberg/issues/56154) [#60708](https://github.com/WordPress/gutenberg/issues/60708) [#60709](https://github.com/WordPress/gutenberg/issues/60709)
 - [x] Performance [#63151](https://github.com/WordPress/gutenberg/issues/63151)
 - [x] Unify Layouts config [#67391](https://github.com/WordPress/gutenberg/issues/67391)
 - [x] Visualize page hierarchy in table/list layout (Hierarchy support for dataviews) [#55957](https://github.com/WordPress/gutenberg/issues/55957)
 - [x] Next steps for search and filtering [#55100](https://github.com/WordPress/gutenberg/issues/55100)
 - [x] General polish [#60847](https://github.com/WordPress/gutenberg/issues/60847) [#59832](https://github.com/WordPress/gutenberg/issues/59832) [#65234](https://github.com/WordPress/gutenberg/issues/65234) [#63719](https://github.com/WordPress/gutenberg/issues/63719) [#65086)(https://github.com/WordPress/gutenberg/issues/65086)

## Next Steps

 - [x] Implement a group by field. [#57967](https://github.com/WordPress/gutenberg/issues/57967)
 - [x] Persisting view configs [#57860](https://github.com/WordPress/gutenberg/issues/57860) [#57669](https://github.com/WordPress/gutenberg/issues/57669)
 - [ ] Custom views next steps (admin menu placement, visibility, persistence) [#56205](https://github.com/WordPress/gutenberg/issues/56205)
 - [ ] Allow injecting extra configuration options for view: For instance the possibility to switch the rendering mode (template-locked/post-only)

## Lower priority
 - [ ] Kanban Layout [#57024](https://github.com/WordPress/gutenberg/issues/57024)
 - [ ] Custom field creation, maybe ? [#55102](https://github.com/WordPress/gutenberg/issues/55102)
 - [ ] Table summary footer [#57270](https://github.com/WordPress/gutenberg/issues/57270)
 - [ ] Support aspect ratio configuration [#60891](https://github.com/WordPress/gutenberg/issues/60891)
 - [ ] Support right click [#59608](https://github.com/WordPress/gutenberg/issues/59608)
 - [x] Support column alignments [#66623](https://github.com/WordPress/gutenberg/issues/66623)
 - [ ] Support pinning columns [#64679](https://github.com/WordPress/gutenberg/issues/64679)
 - [ ] Contextual order control [#65831](https://github.com/WordPress/gutenberg/issues/65831)
 - [ ] Better keyboard accessibility [#56330](https://github.com/WordPress/gutenberg/issues/56330) [#56328 ](https://github.com/WordPress/gutenberg/issues/56328 )

**Comment by dannyreaktiv** [comment]
Is it possible to use the `Table` component outside of the Site Editor?

**Comment by youknowriad** [comment]
@dannyreaktiv Right now the component is named `DataViews` and the goal is to be able to use anywhere. Some part of the current project are specific to the site editors, others are specific to some CPTs... but the underlying component is supposed to be usable anywhere. It's still an internal component at the moment though.

**Comment by dannyreaktiv** [comment]
@youknowriad Thanks for info! It's a great looking component that I look forward to using.

**Comment by dannyreaktiv** [comment]
> @dannyreaktiv Right now the component is named `DataViews` and the goal is to be able to use anywhere. Some part of the current project are specific to the site editors, others are specific to some CPTs... but the underlying component is supposed to be usable anywhere. It's still an internal component at the moment though.

I tried to use `DataViews` as part of a custom block that requires post selection, (following the [nice docs](https://github.com/WordPress/gutenberg/blob/d137227a67c43d04a6cba0cfacf8dd7f95cb74d5/packages/edit-site/src/components/dataviews/README.md#L4)) and ended up giving up. I know this isn't the goal of this issue, but I would love to see some docs on how to use the `DataViews` component outside of the Site Editor. Ok. Thank you. Sorry to take up your time and have a great weekend.

**Comment by youknowriad** [comment]
@dannyreaktiv We'll definitely have some docs in place once it's ready to be used more broadly. Right now, it's changing very quickly and it would be hard to get something stable in place.

**Comment by dannyreaktiv** [comment]
> @dannyreaktiv We'll definitely have some docs in place once it's ready to be used more broadly. Right now, it's changing very quickly and it would be hard to get something stable in place.

That's makes sense 😀👍🙌

**Comment by fabiankaegy** [comment]
Hey all 👋 

I'm coming fresh into this and don't yet have a great overview over the current state of things. I'm wondering if the efforts for making the Data views extensible have already been started / whether we already have any documentation somewhere for how one can add additional "Fields" / columns to the views :) 

I want to help with testing and provide early feedback to help shape this but am currently a bit lost with how I can best support the efforts :) Any insights would be greatly appreciated <3

**Comment by fabiankaegy** [comment]
I've started diving in and testing the data views in general a bit and one of the first pieces of feedback I wanted to provide is how it currently doesn't feel very intuitive how to get to them. Having to always click on "Manage all X" at the bottom of the sidebar adds one additional step to the user flow that I don't feel is very necessary. 

![CleanShot 2024-01-13 at 09 02 26@2x](https://github.com/WordPress/gutenberg/assets/20684594/2b0fabe6-e726-4902-a9fe-766a3c688f98)

What if instead the new "List" view including the search could get integrated as that step directly? Currently when the list view (which I really enjoy by the way!) there is a lot of "wasted" empty space in the left dark sidebar. If these two panels could somehow get combined and therefore also get rid of that "Manage All" intermediate step I think that would be a great enhancement.

![CleanShot 2024-01-13 at 09 02 13@2x](https://github.com/WordPress/gutenberg/assets/20684594/e67157d0-4a10-45ba-823b-60a260e2ddcc)

**Comment by oandregal** [comment]
@fabiankaegy I've added a "Current status" section to summarize where we're at with this effort for all the pages that uses it (Pages, Patterns, Templates).

The interaction you mention is how Pages already works. Not everything is in the “final state”, we've been making progress and testing a few things. The design mockups in the issues linked can provide more info about the target we are aiming for.

**Comment by fabiankaegy** [comment]
Hey @oandregal 👋 

Thanks for those insights. I can confirm I had made a mistake in my testing :) the way it works now for pages is exactly what I was hoping for :) 


What I didn't see answered in the current state is the status of extensibility at this point. Can you say anything regarding that?

**Comment by oandregal** [comment]
> What I didn't see answered in the current state is the status of extensibility at this point. Can you say anything regarding that?

The consumer (Pages, Templates, Patterns) defines the fields and view configuration. There's [API docs](https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/README.md). Is this helpful? If it's not what you are looking for, could you describe your use case?

**Comment by fabiankaegy** [comment]
It is great to see those docs. Thanks for linking them here also. But it isn't really what I am looking for. 

On the old admin tables it was possible to add your own fields to any lists. 

![CleanShot 2024-01-15 at 15 23 46@2x](https://github.com/WordPress/gutenberg/assets/20684594/d55a0129-bdc6-4b66-849d-cb6be84fd930)

Like called out here: https://github.com/WordPress/gutenberg/issues/53233#issuecomment-1665263796 this is very important for the lists of Custom Post Types. But it also is something that is needed for any other standard lists such as Pages, where plugins can add their own columns to these tables that users can then configure using the "Fields" setting. (What used to be available in the "Screen Options")

The ability to showcase custom info in custom columns, regardless of whether it's post meta, custom taxonomies or other data is a crucial customization that we do on most projects. So I'm curious if that is something that has already been considered for these new DataViews as they appear to eventually become the replacement for the old lists (Which I am very very excited about by the way :))

**Comment by oandregal** [comment]
Yeah, that kind of extensibility will be possible. The current focus is getting the APIs (dataviews, fields, view config) in a good state while replacing the existing screens. After that, look at extensibility.

**Comment by fabiankaegy** [comment]
Cool thanks for that confirmation :) I mainly wanted to ensure that I hadn't missed anything for this initial current call for testing where I could have already provided feedback on the expensibilty :)

**Comment by youknowriad** [comment]
Thanks @fabiankaegy for chiming in and for the help and suggestions. I second what @oandregal said. We want this to be as extensible as possible:

 - Registering new fields.
 - Registering new layouts.
 - Registering new field types.
 - Defining the default views and fields.
 - Allowing third-party developers to build their own dataviews powered pages.

That said, it's a big API and we also know that we want to offer some features in these views:

 - Inline editing.
 - editing the fields and views in the UI.

Some of these features might have an impact on the APIs and causing challenges for the backwards compatibility story. For these reasons, I think we want to take a very "secure"/careful approach, initially we'll keep most of these APIs private as we build the features we want in Core to minimize the breaking changes. Once we're confident that the APIs can support all the features that we want, we can start opening them publicly.

One way for third-party developers to start experimenting with these today is to use the `@wordpress/dataviews` npm package as an npm dependency to power their own pages.

I hope this helps.

**Comment by fabiankaegy** [comment]
Hey @youknowriad 👋

Yeah that is very helpful. And it totally makes sense to approach it in that way. 

What I think would help to boost confidence is some kind of roadmap / commitment that defines when this kind of extensibility should be thought about and introduced. 

I agree that it doesn't need there from day one in the new data views inside the site editor. 

But if we think about the larger admin redesign piece I would consider it a requirement before replacing the old pages admin list with the new api.

**Comment by youknowriad** [comment]
> But if we think about the larger admin redesign piece I would consider it a requirement before replacing the old pages admin list with the new api.

I totally agree with this. Extensibility is at the heart of what we're building, we could have designed all these pages in an adhoc way with a simple fixed design, it would have been actually much easier and faster but the WP Admin redesign aspect and the fact that plugins need to be able to modify this in all sorts of way pushed us to think about this in a more extensibility-first way: declarative fields, declarative field types (filters), declarative actions, declarative layouts...

So it seems we're all in agreement here 👍

**Comment by oandregal** [comment]
The PR description has now a section listing the current status for this initiative.

> ### WordPress 6.5 
> 
> Stable:
> 
> - "Site Editor > Templates > Manage all templates": table & grid layouts available.
> - "Site Editor > Template parts > Manage all template parts": table & grid layouts available.
> - "Site Editor > Patterns": grid layout available.
> - "Side Editor > Pages": table & grid layouts available.
> 
> Experimental (enable the "new admin views" experiment):
>
> - "Site Editor > Templates > Manage all templates": table, grid, and list layouts available.
> - "Site Editor > Template Parts > Manage all template parts": table, grid, and list layouts available.
> - "Site Editor > Pages > Manage all pages": table, grid, and list layouts available, as well as custom views (user can save and edit their own views).

**Comment by youknowriad** [comment]
I think all the required items for 6.5 are now merged. Let's remove this one out of the 6.5 board.

**Comment by oandregal** [comment]
For the 6.6 iteration, the whole of the activity is happening elsewhere.

For example, https://github.com/WordPress/gutenberg/issues/59659 tracks migrating all index pages to use DataViews (which was listed here in the 6.5 cycle), tracking tasks for components (filters, bulk actions, etc.), and other related efforts, such as https://github.com/WordPress/gutenberg/issues/59689.

I've updated this issue description for 6.6 to reflect the major areas of the DataViews **component**, leaving any other thing out, as they already have their own tracking issue.

**Comment by oandregal** [comment]
I've updated a bit the issue description pointing to other places where we're tracking goals for WordPress 6.8 and adding a section about "theming", that groups together related issues:

![Image](https://github.com/user-attachments/assets/1525545c-6373-4cee-88b5-071d1a9aba65)

**Comment by youknowriad** [comment]
Hey folks! this issue was becoming a bit hard to follow and scan. I did some triage and updated the issue. I tried to keep it contained and easy to scan and help people understand the priorities... Let me know if you think I've missed anything.

**Comment by jameskoster** [comment]
A couple to consider for low prio:

* https://github.com/WordPress/gutenberg/pull/65086
* https://github.com/WordPress/gutenberg/issues/64679
* https://github.com/WordPress/gutenberg/issues/65831

**Comment by youknowriad** [comment]
Thanks @jameskoster I added them to the list.

**Comment by aaronjorbin** [comment]
[An old core ticket suggested making the bulk edit and quick edit have the same UI](https://core.trac.wordpress.org/ticket/31638).  While Dataviews don't have bulk edit or quick edit right now, I think this is a suggestion worth seriusly considering.

**Comment by youknowriad** [comment]
@aaronjorbin you can enable the quick edit experiment in Gutenberg and both of these are supported with the same UI. There's still some work to make it fully stable though.

**Comment by youknowriad** [comment]
@oandregal what do you think about closing this issue? Do you think it's still useful to keep this one up?

**Comment by oandregal** [comment]
@youknowriad I just took a look, and this one is showing its age. I've been more focused on maintaining the [iteration issue](https://github.com/WordPress/gutenberg/issues/70936) up to date to current work. Let's close this one.
